service/collector/orphan_secret.go

```diff
@@ -0,0 +1,104 @@
+package collector
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/giantswarm/chart-operator/pkg/label"
+	"github.com/giantswarm/chart-operator/service/controller/chart/v1/key"
+)
+
+var (
+	orphanSecretDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, "secret", "orphan"),
+		"Secrets without a chart CR.",
+		[]string{},
+		nil,
+	)
+)
+
+type OrphanSecretConfig struct {
+	G8sClient versioned.Interface
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+}
+
+type OrphanSecret struct {
+	g8sClient versioned.Interface
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+}
+
+func NewOrphanSecret(config OrphanSecretConfig) (*OrphanSecret, error) {
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	oc := &OrphanSecret{
+		g8sClient: config.G8sClient,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return oc, nil
+}
+
+func (oc *OrphanSecret) Collect(ch chan<- prometheus.Metric) error {
+	charts, err := oc.g8sClient.ApplicationV1alpha1().Charts("").List(metav1.ListOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	desiredSecrets := make(map[[2]string]bool)
+
+	for _, chart := range charts.Items {
+		desiredSecrets[[2]string{key.SecretNamespace(chart), key.SecretName(chart)}] = true
```

Comment:
cool I did not know it was possible in go :)

Reply:
This was magic from Mr Pawel.